### PR TITLE
Add USB MSD and EMMC fixes to ChibiOS 21.11.3

### PR DESCRIPTION
**Pull request by bugobliterator**
This pr brings USB Mass Storage support. Also includes fixes to ChibiOS EMMC path.

### CODE REVIEW

File: os/hal/include/hal_usb_msd.h [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+/**
+ * @file    hal_usb_msd.h
+ * @brief   USM mass storage device driver macros and structures.
```
Comment (andyp1per) [inline os/hal/include/hal_usb_msd.h:19]: USB rather than USM?

File: os/hal/hal.mk [hunk 2]
```diff
@@ -95,6 +95,11 @@ endif
 ifneq ($(findstring HAL_USE_WSPI TRUE,$(HALCONF)),)
 HALSRC += $(CHIBIOS)/os/hal/src/hal_wspi.c
 endif
+ifneq ($(findstring HAL_USE_USB_MSD TRUE,$(HALCONF)),)
```
Comment (andyp1per) [inline os/hal/hal.mk:98]: Are you sure this works? I was having problems yesterday

File: os/hal/ports/STM32/LLD/OTGv1/stm32_otg.h [hunk 3]
```diff
@@ -33,7 +33,7 @@
 /**
  * @brief   OTG_HS FIFO memory size in words.
  */
-#define STM32_OTG2_FIFO_MEM_SIZE        1024
+#define STM32_OTG2_FIFO_MEM_SIZE        4096
```
Comment (andyp1per) [inline os/hal/ports/STM32/LLD/OTGv1/stm32_otg.h:36]: This is a fairly significant increase, does it need to be protected with a #define to conserve memory?

File: os/hal/src/hal_sdc.c [hunk 4]
```diff
@@ -690,19 +693,13 @@ bool sdcConnect(SDCDriver *sdcp) {
...
-  sdc_lld_set_data_clk(sdcp, clk);
-
   /* Reads extended CSD if needed and possible.*/
   if (SDC_MODE_CARDTYPE_MMC == (sdcp->cardmode & SDC_MODE_CARDTYPE_MASK)) {
 
     /* The card is a MMC, checking if it is a large device.*/
     if (_mmcsd_get_slice(sdcp->csd, MMCSD_CSD_MMC_CSD_STRUCTURE_SLICE) > 1U) {
       uint8_t *ext_csd = sdcp->buf;
-
+      osalThreadSleepMilliseconds(10);
```
Comment (andyp1per) [inline os/hal/src/hal_sdc.c:702]: The newer driver code does these sleeps in a slightly different way - do you think we need to copy?

File: os/various/ramdisk.c [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+/**
+ * @file    ramdisk.c
+ * @brief   Virtual block devise driver source.
```
Comment (andyp1per) [inline os/various/ramdisk.c:19]: device

File: os/various/ramdisk.h [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+/**
+ * @file    ramdisk.h
+ * @brief   Virtual block devise driver header.
```
Comment (andyp1per) [inline os/various/ramdisk.h:19]: device

**Review by andyp1per — COMMENTED** [review]
I think you will need @tridge to review as well

File: os/hal/hal.mk (hunk 2 above)
Comment (bugobliterator) [reply os/hal/hal.mk:98]: Its working fine last I tested

File: os/hal/src/hal_sdc.c (hunk 4 above)
Comment (bugobliterator) [reply os/hal/src/hal_sdc.c:702]: can you share an example.

File: os/hal/ports/STM32/LLD/OTGv1/stm32_otg.h (hunk 3 above)
Comment (bugobliterator) [reply os/hal/ports/STM32/LLD/OTGv1/stm32_otg.h:36]: This is for OTG HS only, doesn't impact any existing hardware, this is needed for better throughput on MSD

**Comment by andyp1per** [comment]
Can you rebase @bugobliterator ?

**Comment by bugobliterator** [comment]
> Can you rebase @bugobliterator ?

done

**Comment by andyp1per** [comment]
Sorry, @bugobliterator needs rebasing again

**Comment by bugobliterator** [comment]
> Sorry, @bugobliterator needs rebasing again

done @andyp1per

**Review by andyp1per — APPROVED** [review]

**Review by andyp1per — APPROVED** [review]

**Pull request merged**